File: conformance.go
package xtxt

// Canonical converts a document to the normalised shape used by the
// conformance suite. Every field is present with an explicit default, so an
// implementation cannot pass by omitting something the reference emits — and
// two implementations cannot appear to differ merely because one of them
// leaves empty fields out of its JSON.
func Canonical(doc *Document) map[string]any {
	nodes := make([]any, 0, len(doc.Nodes))
	for _, n := range doc.Nodes {
		args := make([]any, 0, len(n.Args))
		for _, a := range n.Args {
			args = append(args, map[string]any{"key": a.Key, "value": a.Value})
		}
		items := make([]any, 0, len(n.Items))
		for _, it := range n.Items {
			var checked any
			if it.Checked != nil {
				checked = *it.Checked
			}
			items = append(items, map[string]any{
				"text": it.Text, "ordered": it.Ordered, "checked": checked,
			})
		}
		nodes = append(nodes, map[string]any{
			"kind":  string(n.Kind),
			"name":  n.Name,
			"level": n.Level,
			"text":  n.Text,
			"args":  args,
			"items": items,
			"line":  n.Line,
		})
	}
	return map[string]any{"version": doc.Version, "nodes": nodes}
}

// CanonicalIssues normalises diagnostics for the conformance suite. Messages
// are implementation wording and are deliberately excluded: what every
// implementation must agree on is *that* a line is an error, not how it phrases
// the complaint.
func CanonicalIssues(issues []Issue) []any {
	out := make([]any, 0, len(issues))
	for _, i := range SortIssues(issues) {
		out = append(out, map[string]any{
			"severity": string(i.Severity),
			"line":     i.Line,
		})
	}
	return out
}

File: sdk/rust/tests/conformance.rs
//! The Rust SDK is checked against the same fixtures as the Go reference.
//!
//! `canonical` lives here rather than in the crate so that the library itself
//! stays dependency-free: only the test needs to speak JSON.

use serde_json::{json, Value};
use std::fs;
use std::path::{Path, PathBuf};

fn cases_dir() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("../../conformance/cases")
        .canonicalize()
        .expect("conformance cases directory")
}

/// The normalised shape every implementation must produce. Every field is
/// present with an explicit default, so an implementation cannot pass by
/// omitting something the reference emits.
fn canonical(doc: &xtxt::Document) -> Value {
    let nodes: Vec<Value> = doc
        .nodes
        .iter()
        .map(|n| {
            let args: Vec<Value> = n
                .args
                .iter()
                .map(|a| json!({ "key": a.key, "value": a.value }))
                .collect();
            let items: Vec<Value> = n
                .items
                .iter()
                .map(|it| {
                    json!({
                        "text": it.text,
                        "ordered": it.ordered,
                        "checked": match it.checked {
                            Some(v) => Value::Bool(v),
                            None => Value::Null,
                        },
                    })
                })
                .collect();
            json!({
                "kind": n.kind.as_str(),
                "name": n.name,
                "level": n.level,
                "text": n.text,
                "args": args,
                "items": items,
                "line": n.line,
            })
        })
        .collect();
    json!({ "version": doc.version, "nodes": nodes })
}

fn canonical_issues(issues: &[xtxt::Issue]) -> Value {
    Value::Array(
        xtxt::sort_issues(issues)
            .iter()
            .map(|i| json!({ "severity": i.severity.as_str(), "line": i.line }))
            .collect(),
    )
}

#[test]
fn matches_the_reference_implementation() {
    let dir = cases_dir();
    let mut inputs: Vec<PathBuf> = fs::read_dir(&dir)
        .expect("read cases")
        .filter_map(|e| e.ok().map(|e| e.path()))
        .filter(|p| p.extension().map(|x| x == "xtxt").unwrap_or(false))
        .collect();
    inputs.sort();
    assert!(!inputs.is_empty(), "no conformance cases found in {dir:?}");

    let mut failures = Vec::new();
    for input in &inputs {
        let src = fs::read_to_string(input).expect("read case");
        let expected: Value = serde_json::from_str(
            &fs::read_to_string(input.with_extension("json")).expect("read expectation"),
        )
        .expect("parse expectation");

        let res = xtxt::parse(&src);
        let mut issues = res.issues.clone();
        issues.extend(xtxt::validate(&res.doc));

        let name = input.file_name().unwrap().to_string_lossy().to_string();
        if canonical(&res.doc) != expected["ast"] {
            failures.push(format!(
                "{name}: AST mismatch\n  got:  {}\n  want: {}",
                canonical(&res.doc),
                expected["ast"]
            ));
        }
        if canonical_issues(&issues) != expected["issues"] {
            failures.push(format!(
                "{name}: issues mismatch\n  got:  {}\n  want: {}",
                canonical_issues(&issues),
                expected["issues"]
            ));
        }
    }
    assert!(
        failures.is_empty(),
        "{} of {} cases failed:\n{}",
        failures.len(),
        inputs.len(),
        failures.join("\n")
    );
}
